File: auto-vps-agent/agent.py
#!/usr/bin/env python3
"""
CK-NEXUS Auto-VPS Agent Engine
Scans free VPS providers, analyzes with AI, auto-registers
"""

import os
import sys
import json
import time
import sqlite3
import hashlib
import subprocess
import urllib.request
import urllib.error
from typing import Dict, Any, List

SD_PATH = "/workspace/ck-nexus"
CONFIG_PATH = "/root/.ck-nexus/config.json"

# Free VPS Providers
FREE_VPS_PROVIDERS = [
    {"name": "Oracle Cloud", "url": "https://cloud.oracle.com/free_signup", "type": "always-free", "specs": "4 OCPU/24GB", "days": 365, "needs_card": False},
    {"name": "GratisVPS", "url": "https://gratisvps.net/cvps", "type": "free", "specs": "2GB/2CPU", "days": -1, "needs_card": False},
    {"name": "SolusVM", "url": "https://www.solusvm.com/free-trial", "type": "trial", "specs": "Sandbox", "days": 30, "needs_card": False},
    {"name": "Google Cloud", "url": "https://console.cloud.google.com/billing/free", "type": "free-tier", "specs": "$300 credit", "days": 90, "needs_card": True},
    {"name": "Kamatera", "url": "https://www.kamatera.com/free-trial/", "type": "trial", "specs": "4CPU/8GB", "days": 30, "needs_card": False},
]


def load_config():
    with open(CONFIG_PATH) as f:
        return json.load(f)


async def analyze_with_ai(text: str) -> Dict[str, Any]:
    """Use OpenRouter AI to analyze VPS offers"""
    config = load_config()
    openrouter_key = config.get("openrouter", {}).get("key", "")

    prompt = f"""คุณคือ AI นักวิเคราะห์ VPS ฟรี
วิเคราะห์ข้อความนี้: "{text[:1500]}"

ตอบกลับเป็น JSON เท่านั้น:
{{"isFreeVPS": true/false, "trustScore": 1-10, "reason": "เหตุผลสั้นๆ", "recommended": true/false}}"""

    try:
        req = urllib.request.Request(
            "https://openrouter.ai/api/v1/chat/completions",
            data=json.dumps({
                "model": "mistralai/mistral-7b-instruct:free",
                "messages": [{"role": "user", "content": prompt}],
                "max_tokens": 300,
                "temperature": 0.3
            }).encode(),
            headers={
                "Authorization": f"Bearer {openrouter_key}",
                "Content-Type": "application/json"
            }
        )
        with urllib.request.urlopen(req, timeout=15) as resp:
            data = json.loads(resp.read())
            content = data["choices"][0]["message"]["content"]
            import re
            match = re.search(r'\{[\s\S]*\}', content)
            if match:
                return json.loads(match.group())
    except Exception as e:
        print(f"  ⚠️ AI error: {e}")
    return {"isFreeVPS": False, "trustScore": 0, "reason": "AI failed", "recommended": False}


def save_to_db(provider_name: str, ip: str = "", password: str = "", status: str = "AWAITING"):
    db_path = os.path.join(SD_PATH, "nexus_system_sd.db")
    with sqlite3.connect(db_path) as conn:
        conn.execute("""
            UPDATE autonomous_vps_servers 
            SET vps_ip=?, vps_password=?, status=?, notes=?, timestamp=datetime('now')
            WHERE provider_name LIKE ?
        """, (ip, password, status, f"SSH_IP:{ip}|SSH_PASSWORD:{password}", f"%{provider_name}%"))
        conn.commit()
    print(f"  💾 Saved to DB: {provider_name}")


def scan_providers():
    """Scan all free VPS providers"""
    print("╔══════════════════════════════════════════════╗")
    print("║  🤖 CK-NEXUS AUTO-VPS AGENT ENGINE          ║")
    print("╠══════════════════════════════════════════════╣")
    print("║  🧠 AI: Groq Llama 3.3 70B                  ║")
    print("║  🎯 Target: Free VPS Providers               ║")
    print("╚══════════════════════════════════════════════╝")
    print()

    results = []

    for provider in FREE_VPS_PROVIDERS:
        print(f"\n🔍 Scanning: {provider['name']}")
        print(f"   URL: {provider['url']}")
        print(f"   Type: {provider['type']} | Days: {provider['days']} | Card: {'Yes' if provider['needs_card'] else 'No'}")

        # Try to fetch page
        try:
            req = urllib.request.Request(provider['url'], headers={
                'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36'
            })
            with urllib.request.urlopen(req, timeout=10) as resp:
                content = resp.read().decode(errors='ignore')[:3000]

                # Analyze with AI
                import asyncio
                analysis = asyncio.run(analyze_with_ai(content))

                if analysis.get('recommended', False) or analysis.get('trustScore', 0) >= 7:
                    print(f"   ✅ AI Approved: {analysis.get('reason', '')} (Score: {analysis.get('trustScore', 0)})")
                    save_to_db(provider['name'], status="READY_FOR_REGISTRATION")
                    results.append({"provider": provider['name'], "status": "APPROVED", "analysis": analysis})
                else:
                    print(f"   ❌ AI Rejected: {analysis.get('reason', '')}")
                    results.append({"provider": provider['name'], "status": "REJECTED", "analysis": analysis})
        except Exception as e:
            print(f"   ⚠️ Fetch error: {e}")
            results.append({"provider": provider['name'], "status": "ERROR", "error": str(e)})

    return results


def generate_registration_scripts():
    """Generate registration scripts for each provider"""
    print("\n📋 Generating registration scripts...")

    # Oracle Cloud
    oracle_script = f"""#!/bin/bash
# Oracle Cloud Free Tier Registration
# No credit card required for Always Free tier

echo "=== Oracle Cloud Free Tier Setup ==="
echo ""
echo "1. Go to: https://cloud.oracle.com"
echo "2. Click 'Sign Up for Free Tier'"
echo "3. Use email: iwepnewqviay800@gmail.com"
echo "4. Complete registration"
echo "5. Create Compute Instance:"
echo "   - Image: Ubuntu 22.04"
echo "   - Shape: Ampere ARM (4 OCPU/24GB)"
echo ""
echo "6. SSH Public Key:"
cat /root/.ssh/oracle_key.pub
echo ""
echo "7. After creating instance, get Public IP"
echo "8. Connect: ssh -i /root/.ssh/oracle_key ubuntu@<IP>"
echo "9. Set password: sudo passwd ubuntu"
echo ""
echo "10. Send to Telegram: add vps oracle <IP> <Password>"
"""
    with open(os.path.join(SD_PATH, "oracle_register.sh"), "w") as f:
        f.write(oracle_script)
    os.chmod(os.path.join(SD_PATH, "oracle_register.sh"), 0o755)

    # GratisVPS
    gratis_script = """#!/bin/bash
# GratisVPS Registration - No Credit Card Required

echo "=== GratisVPS Setup ==="
echo ""
echo "1. Go to: https://gratisvps.net/cvps"
echo "2. Fill in email: iwepnewqviay800@gmail.com"
echo "3. Create password: CK-Nexus-2026!"
echo "4. Complete registration"
echo "5. Deploy free VPS instance"
echo ""
echo "6. Send to Telegram: add vps gratisvps <IP> <Password>"
"""
    with open(os.path.join(SD_PATH, "gratisvps_register.sh"), "w") as f:
        f.write(gratis_script)
    os.chmod(os.path.join(SD_PATH, "gratisvps_register.sh"), 0o755)

    # SolusVM
    solusvm_script = """#!/bin/bash
# SolusVM Developer Trial - No Credit Card Required

echo "=== SolusVM Dev Trial Setup ==="
echo ""
echo "1. Go to: https://www.solusvm.com/free-trial"
echo "2. Fill in registration form"
echo "3. Use email: iwepnewqviay800@gmail.com"
echo "4. Complete registration"
echo "5. Deploy trial VPS instance"
echo ""
echo "6. Send to Telegram: add vps solusvm <IP> <Password>"
"""
    with open(os.path.join(SD_PATH, "solusvm_register.sh"), "w") as f:
        f.write(solusvm_script)
    os.chmod(os.path.join(SD_PATH, "solusvm_register.sh"), 0o755)

    print("  ✅ oracle_register.sh")
    print("  ✅ gratisvps_register.sh")
    print("  ✅ solusvm_register.sh")


def main():
    if len(sys.argv) > 1:
        cmd = sys.argv[1]

        if cmd == "scan":
            results = scan_providers()
            print("\n" + "=" * 50)
            print("📊 SCAN RESULTS:")
            print("=" * 50)
            for r in results:
                status = "✅" if r["status"] == "APPROVED" else "❌"
                print(f"  {status} {r['provider']}: {r['status']}")

        elif cmd == "scripts":
            generate_registration_scripts()

        elif cmd == "configure":
            if len(sys.argv) < 4:
                print("Usage: python3 agent.py configure <IP> <PASSWORD> [provider]")
                sys.exit(1)
            ip = sys.argv[2]
            password = sys.argv[3]
            provider = sys.argv[4] if len(sys.argv) > 4 else "oracle"
            save_to_db(provider, ip, password, "ACTIVE")
            print(f"✅ VPS {provider} configured at {ip}")

        elif cmd == "all":
            generate_registration_scripts()
            results = scan_providers()
            print("\n✅ Auto-VPS Agent complete!")
            print("📋 Registration scripts saved to:", SD_PATH)

        else:
            print("Commands: scan, scripts, configure, all")
    else:
        generate_registration_scripts()
        results = scan_providers()
        print("\n✅ Auto-VPS Agent complete!")


if __name__ == "__main__":
    main()

File: auto-vps-agent/agent.js
const { PlaywrightCrawler } = require('crawlee');
const { chromium } = require('playwright');
const fs = require('fs');
const path = require('path');
const https = require('https');
const http = require('http');

// Load config
const config = JSON.parse(fs.readFileSync('/root/.ck-nexus/config.json', 'utf8'));
const GROQ_KEY = config.groq?.key || config.GROQ_API_KEY || '';
const OPENROUTER_KEY = config.openrouter?.key || config.OPENROUTER_API_KEY || '';
const SD_PATH = '/workspace/ck-nexus';

// VPS Providers that offer free tiers
const FREE_VPS_PROVIDERS = [
    {
        name: 'Oracle Cloud',
        url: 'https://cloud.oracle.com',
        type: 'always-free',
        specs: '4 OCPU/24GB RAM',
        days: 365
    },
    {
        name: 'GratisVPS',
        url: 'https://gratisvps.net/cvps',
        type: 'free',
        specs: '2GB RAM/2 CPU',
        days: -1
    },
    {
        name: 'SolusVM',
        url: 'https://www.solusvm.com/free-trial',
        type: 'trial',
        specs: 'Sandbox',
        days: 30
    },
    {
        name: 'Google Cloud',
        url: 'https://console.cloud.google.com/billing/free',
        type: 'free-tier',
        specs: '$300 credit/e2-micro',
        days: 90
    }
];

// AI Analysis using Groq
async function analyzeWithAI(text) {
    try {
        const prompt = `คุณคือ AI นักวิเคราะห์ระบบคลาวด์และ VPS ฟรี
วิเคราะห์ข้อความนี้: "${text}"

ตอบกลับเป็น JSON เท่านั้น:
{
  "isFreeVPS": true/false,
  "trustScore": 1-10,
  "reason": "เหตุผลสั้นๆ ภาษาไทย",
  "recommended": true/false
}`;

        const response = await fetch('https://api.groq.com/openai/v1/chat/completions', {
            method: 'POST',
            headers: {
                'Authorization': `Bearer ${GROQ_KEY}`,
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                model: 'llama-3.3-70b-versatile',
                messages: [{ role: 'user', content: prompt }],
                max_tokens: 500,
                temperature: 0.3
            })
        });

        if (response.ok) {
            const data = await response.json();
            const content = data.choices[0].message.content;
            const jsonMatch = content.match(/\{[\s\S]*\}/);
            if (jsonMatch) {
                return JSON.parse(jsonMatch[0]);
            }
        }
    } catch (e) {
        console.log(`⚠️ AI analysis error: ${e.message}`);
    }
    return { isFreeVPS: false, trustScore: 0, reason: 'AI analysis failed', recommended: false };
}

// Save VPS to database
function saveVPSToDB(provider, ip, password) {
    const dbPath = path.join(SD_PATH, 'nexus_system_sd.db');
    const sqlite3 = require('sqlite3').verbose();
    const db = new sqlite3.Database(dbPath);

    db.run(`UPDATE autonomous_vps_servers 
            SET vps_ip=?, vps_password=?, status='ACTIVE', 
                notes=?, timestamp=datetime('now')
            WHERE provider_name LIKE ?`,
        [ip, password, `SSH_IP:${ip}|SSH_PASSWORD:${password}`, `%${provider}%`],
        function(err) {
            if (err) console.log(`⚠️ DB error: ${err.message}`);
            else console.log(`💾 Saved ${provider} to database`);
            db.close();
        }
    );
}

// Main crawler
const crawler = new PlaywrightCrawler({
    launchContext: {
        launchOptions: {
            headless: true,
            args: ['--no-sandbox', '--disable-setuid-sandbox']
        }
    },
    maxRequestsPerCrawl: 20,
    navigationTimeoutSecs: 30,

    async requestHandler({ page, request, log }) {
        log.info(`🔍 Checking: ${request.url}`);

        try {
            await page.waitForLoadState('domcontentloaded', { timeout: 15000 });

            // Analyze page content
            const pageText = await page.textContent('body');
            const analysis = await analyzeWithAI(pageText.substring(0, 2000));

            if (analysis.isFreeVPS && analysis.trustScore >= 7) {
                log.info(`✅ AI Approved: ${analysis.reason} (Score: ${analysis.trustScore})`);

                // Take screenshot
                const screenshotPath = path.join(SD_PATH, `vps_${Date.now()}.png`);
                await page.screenshot({ path: screenshotPath, fullPage: false });
                log.info(`📸 Screenshot saved: ${screenshotPath}`);

                // Try to find signup button
                const signupBtn = await page.$('a[href*="signup"], a[href*="register"], a[href*="create"], button:has-text("Sign Up"), button:has-text("Register"), button:has-text("Create")');
                if (signupBtn) {
                    log.info(`🚀 Found signup button, clicking...`);
                    await signupBtn.click();
                    await page.waitForLoadState('domcontentloaded', { timeout: 15000 });

                    // Capture form fields
                    const formFields = await page.$$eval('input[type="email"], input[type="text"], input[type="password"]', inputs =>
                        inputs.map(i => ({ type: i.type, name: i.name || i.id, placeholder: i.placeholder }))
                    );
                    log.info(`📝 Found ${formFields.length} form fields: ${JSON.stringify(formFields)}`);

                    // Save form structure
                    fs.writeFileSync(path.join(SD_PATH, `vps_form_${Date.now()}.json`), JSON.stringify({
                        url: request.url,
                        fields: formFields,
                        screenshot: screenshotPath,
                        analysis: analysis
                    }, null, 2));
                }
            } else {
                log.info(`❌ AI Rejected: ${analysis.reason}`);
            }
        } catch (e) {
            log.info(`⚠️ Error processing page: ${e.message}`);
        }
    }
});

// Start the agent
async function main() {
    console.log('╔══════════════════════════════════════════════╗');
    console.log('║  🤖 CK-NEXUS AUTO-VPS AGENT ENGINE          ║');
    console.log('╠══════════════════════════════════════════════╣');
    console.log('║  🧠 AI: Groq Llama 3.3 70B                  ║');
    console.log('║  🕵️ Bot: Crawlee + Playwright                ║');
    console.log('║  🎯 Target: Free VPS Providers               ║');
    console.log('╚══════════════════════════════════════════════╝');
    console.log('');

    // Scan free VPS sources
    const startUrls = [
        'https://gratisvps.net/cvps',
        'https://www.solusvm.com/free-trial',
        'https://lowendtalk.com'
    ];

    console.log(`🎯 Scanning ${startUrls.length} sources for free VPS...`);

    try {
        await crawler.run(startUrls);
    } catch (e) {
        console.log(`⚠️ Crawler stopped: ${e.message}`);
    }

    console.log('\n✅ Scan complete! Check screenshots in', SD_PATH);
}

main().catch(console.error);
